docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,253 @@
+    - Class C direct downstream messages sent from the portal could result in a connection switch if
+    the handling LNS changed in the meantime but we consider this as out of scope.
+
+Version, LNS discovery and CUPS update endpoints are not affected.
+
+## Possible solution
+
+The main idea is to give the current connection holder as indicated from the Function, the edge to
+continue processing messages for this device. That gateway, will have 0 impact on performance.
+
+### Handling of background tasks
+
+When we create a (singleton) instance of the NetworkServer.LoRaDeviceCache we start a background periodical
+check to ensure the device twins for the DevEuis we encountered are kept fresh.
+
+- If we were previously the losing LNS for the DevEui that needs refresh, we could adjust the
+  LastSeen property but not actually refresh the entry
+  - this way we won't try to read the twin that could result in a connection switch
+  - when the next data message comes and we check the cache
+    - if entry is in cache, we have a stale twin in the cache - how should it be solved ❔
+    - if entry not in cache, see handling in [the section on LoRaDevice not in cache](#data-flow-loradevice-not-in-loradevicecache)
+  
+Alternatives
+
+- Remove entry completely from the cache which is the same as [the section on LoRaDevice not in
+  cache](#data-flow-loradevice-not-in-loradevicecache)
+- Delay and then call the Function to check whether we are still the losing LNS
+  - Not really a background task anymore
+- Do nothing and accept there will be a potential connection switching periodically (currently every
+  2 days)
+- Remove the background refresh feature all-together.
+
+### Handling of Join requests
+
+Join requests in isolation currently do not have the connection stealing issue, as they already rely on the
+Function for the DevNonce check. If the current LNS is not the preferred gateway, it drops the
+message immediately.
+
+❔ We could also consider storing the information that we were the losing LNS locally on the LNS and
+handle "backoff" in the same way as in the Data flow. The advantages of doing that are:
+
+- subsequent first Data (or future Join) requests do not switch the connection
+- same handling as in the Data flow (therefore easier to understand)
+  
+Alternative is to do nothing and accept that a one-off connection stealing on the first Data message can happen.
+
+### Data flow: LoRaDevice not in LoRaDeviceCache
+
+Currently, if LoRaDevice is not in the cache we search on the Function for devices with the specific
+DevAddr then we load their twins.
+
+We could ❔
+
+- Amend DeviceGetter.GetDevice so that it returns info about whether we are the winning gateway
+  - if we are (we already have the connection) continue to load the twin
+  - if we are not the winning LNS, drop the message and mark our selves as the losing gateway
+  - This would require moving the Mic computation on the Function (?)
```

Comment:
I Do not understand why this would change MIC computation, could we not simply do that when after the getdevice

Reply:
I am not sure as well - I discussed this with @danigian and left the note there but don't get it either.

Reply:
After the discussion, we need the Mic computation to see if we can decrypt the message (find out the real DevEui from the DevAddr). This is something we can not do on the Function because we need to send the payload to the cloud.